File: examples/viz/web_viz/static/js/illumination.js
// Applies the scenario sun truth (from the backend map_setup payload) to
// every lighting consumer: the 2D hillshade and the 3D sun light + shadows.
// One truth source keeps what the operator sees consistent with the sun the
// engine uses for sensor glare adjudication.

import { vizState } from './store.js';
import { refreshIlluminationShading } from './scene-geometry.js';
import { updateSceneIllumination } from './scene3d.js';

export function applyIllumination(raw) {
    if (!raw || typeof raw !== 'object') return;
    const ill = vizState.illumination;
    const azimuth = Number(raw.sun_azimuth_deg);
    const elevation = Number(raw.sun_elevation_deg);
    const nextAzimuth = Number.isFinite(azimuth) ? ((azimuth % 360) + 360) % 360 : ill.sunAzimuthDeg;
    const nextElevation = Number.isFinite(elevation) ? Math.max(-90, Math.min(90, elevation)) : ill.sunElevationDeg;
    const changed = nextAzimuth !== ill.sunAzimuthDeg || nextElevation !== ill.sunElevationDeg;
    ill.sunAzimuthDeg = nextAzimuth;
    ill.sunElevationDeg = nextElevation;
    ill.configured = !!raw.configured;
    ill.engineConfirmed = !!raw.engine_confirmed;
    if (changed) {
        updateSceneIllumination();
        refreshIlluminationShading();
        console.log(
            `Illumination applied: sun az ${nextAzimuth.toFixed(1)} deg, el ${nextElevation.toFixed(1)} deg`
            + (ill.engineConfirmed ? ' (engine-confirmed)' : ''),
        );
    }
}

File: examples/viz/runtime/illumination.py
"""Resolve the scenario sun-illumination truth for the viz frontend.

The engine owns the operational sun direction (``IEnvironmentModel``); the
scenario's ``environment.illumination`` block is the configuration that both
the kernel apply path and this viz payload read, so display and sensor-glare
adjudication share one source. When the sim handle exposes the kernel getter
we cross-check that the engine actually holds the same direction.
"""

from __future__ import annotations

import math
from typing import Any

DEFAULT_SUN_AZIMUTH_DEG = 0.0
DEFAULT_SUN_ELEVATION_DEG = 45.0


def sun_vector_from_angles(azimuth_deg: float, elevation_deg: float) -> tuple[float, float, float]:
    """Unit vector toward the sun in ENU (east, north, up).

    NAV azimuth: 0 = North (+Y), clockwise positive toward East (+X).
    Mirrors DefaultEnvironmentModel::get_sun_direction.
    """
    az = math.radians(azimuth_deg)
    el = math.radians(elevation_deg)
    horizontal = math.cos(el)
    return (math.sin(az) * horizontal, math.cos(az) * horizontal, math.sin(el))


def resolve_scenario_illumination(
    scenario_data: dict[str, Any] | None,
    *,
    sim: Any = None,
) -> dict[str, Any]:
    """Build the ``illumination`` block for the viz ``map_setup`` payload."""
    env_cfg = (scenario_data or {}).get("environment", {})
    if not isinstance(env_cfg, dict):
        env_cfg = {}
    raw_cfg = env_cfg.get("illumination", None)
    configured = isinstance(raw_cfg, dict)
    cfg = raw_cfg if configured else {}

    try:
        azimuth = float(cfg.get("sun_azimuth_deg", DEFAULT_SUN_AZIMUTH_DEG))
    except (TypeError, ValueError):
        azimuth = DEFAULT_SUN_AZIMUTH_DEG
    try:
        elevation = float(cfg.get("sun_elevation_deg", DEFAULT_SUN_ELEVATION_DEG))
    except (TypeError, ValueError):
        elevation = DEFAULT_SUN_ELEVATION_DEG

    # Same normalization the engine applies.
    azimuth = azimuth % 360.0
    elevation = max(-90.0, min(90.0, elevation))

    payload: dict[str, Any] = {
        "sun_azimuth_deg": azimuth,
        "sun_elevation_deg": elevation,
        "configured": configured,
        "engine_confirmed": False,
    }

    if sim is not None and hasattr(sim, "get_sun_direction"):
        try:
            engine_vec = tuple(float(v) for v in sim.get_sun_direction())
            expected_vec = sun_vector_from_angles(azimuth, elevation)
            dot = sum(a * b for a, b in zip(engine_vec, expected_vec))
            payload["engine_confirmed"] = bool(dot > 0.99999)
        except Exception:
            pass

    return payload
